### PR TITLE
Add PodHealthyPolicy for PDBs

### PR DESCRIPTION
**Pull request by mortent**
#### What type of PR is this?
This PR implements the `PodHealthyPolicy` for PodDisruptionBudgets as described in https://github.com/kubernetes/enhancements/issues/3017

/kind feature
/kind api-change
/sig apps

#### What this PR does / why we need it:
This allows users to specify how PodDisruptionBudget and the eviction API should handle pods that are `Running` but not `Ready`.

#### Does this PR introduce a user-facing change?

```release-note
Add PodHealthyPolicy for PodDisruptionBudget
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3017-pod-healthy-policy-for-pdb
```

### CODE REVIEW

**Comment by jpbetz** [comment]
cc @deads2k 
/triage accepted

**Review by soltysh — CHANGES_REQUESTED** [review]
You're missing clearing the `PodHealthyPolicy` field when the feature is disabled, similarly to how we do that for example here: 
- for create https://github.com/kubernetes/kubernetes/blob/a83c97941b63e800805378124a5fab08a051e4d4/pkg/registry/batch/job/strategy.go#L96-L102
- for update https://github.com/kubernetes/kubernetes/blob/a83c97941b63e800805378124a5fab08a051e4d4/pkg/registry/batch/job/strategy.go#L132-L134

File: pkg/apis/policy/types.go [hunk 1]
```diff
@@ -42,8 +42,37 @@ type PodDisruptionBudgetSpec struct {
...
 	MaxUnavailable *intstr.IntOrString
+
+	// PodHealthyPolicy defines the criteria for when the disruption controller
+	// should consider a pod to be healthy.
+	// If no policy is specified, the legacy behavior will be used. It means
+	// only pods that are Running and Ready will be considered when the disruption
+	// controller computes "disruptionsAllowed", but all pods in the Running phase
+	// will be subject to the PDB on eviction.
+	// +optional
+	PodHealthyPolicy PodHealthyPolicy
```
Comment (soltysh) [inline pkg/apis/policy/types.go:53]: Add information this is alpha-level field and that feature flag needs to be enabled for it to work: similarly to https://github.com/kubernetes/kubernetes/blob/a83c97941b63e800805378124a5fab08a051e4d4/pkg/apis/batch/types.go#L246-L251

**Review by soltysh — CHANGES_REQUESTED** [review]

**Comment by atiratree** [comment]
hi @mortent, we are building on top of this PR with @ravisantoshgudimetla  -

- https://github.com/kubernetes/kubernetes/pull/113375
- https://github.com/kubernetes/kubernetes/pull/113376

**Pull request closed without merging**